allocator_test.go

```diff
@@ -1,23 +1,239 @@
+	rangeDiskSpaceDemands := [RangeSizeLimit]int64{85, 75, 12, 11, 10, 9, 8, 7, 6, 6}
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	expectedAllocation := map[allocator.RangeID][]allocator.NodeID{
+		0: {2},
+		1: {1},
+		2: {0},
+		3: {0},
+		4: {0},
+		5: {0},
+		6: {0},
+		7: {0},
+		8: {0},
+		9: {0},
+	}
+	require.True(t, status)
+	require.Equal(t, expectedAllocation, allocation)
+}
+
+func TestGIVENCapacityConstraintWithReplicationWHENMultipleAllocationsPossibleTHENAllocationSucceeds(t *testing.T) {
+	const RangeSizeLimit = 5
+	const ReplicationFactor = 3
+	const TestClusterSize = 3
+	clusterCapacities := []int64{70, 80, 90}
+	rangeDiskSpaceDemands := []int64{25, 10, 12, 11, 10}
+	builtCluster := buildCluster(TestClusterSize, clusterCapacities)
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.True(t, status)
+	for _, nodeAssignments := range allocation {
+		require.Equal(t, len(nodeAssignments), ReplicationFactor)
+		require.True(t, isValidNodeAssignment(nodeAssignments, TestClusterSize))
+		require.True(t, isEachReplicaAssignedToDifferentNode(nodeAssignments, TestClusterSize))
+	}
+	require.True(t, nodeCapacityIsRespected(allocation, clusterCapacities, rangeDiskSpaceDemands))
+}
+
+func TestGIVENCapacityConstraintWithReplicationWHENInsufficientResourcesTHENAllocationFails(t *testing.T) {
+	const RangeSizeLimit = 5
+	const ReplicationFactor = 5
+	const TestClusterSize = 3
+	clusterCapacities := []int64{70, 80, 90}
+	rangeDiskSpaceDemands := []int64{25, 10, 12, 11, 10}
+	builtCluster := buildCluster(TestClusterSize, clusterCapacities)
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.False(t, status)
+	require.Nil(t, allocation)
+}
+
+func TestGIVENCapacityConstraintWHENInsufficientNodeCapacityTHENAllocationFails(t *testing.T) {
```

Comment:
Use better names for these tests. For this one, I'd use something like TestInsufficientNodeCapacity